prov/verbs/src/fi_verbs.c

```diff
@@ -729,6 +729,8 @@ static void fi_ibv_fini(void)
 
 VERBS_INI
 {
+	setenv("MLX4_SINGLE_THREADED", "1", 0);
+	setenv("MLX5_SINGLE_THREADED", "1", 0);
```

Comment:
What libibverbs version has this support?  Personally, I don't trust Mellanox to ensure backwards compatibility for these environment variables, and we need to know that the underlying verbs libraries handle this variable.  Btw, verbs is working on adding a standard way for removing locks from the code using 'parent' domains.  If that option is available, we should attempt to use it instead.

Reply:
I know nothing about this update in libibverbs.
I agree it would be good to have this. But these spinlocks are used in mlx4/mlx5 libraries (mlx5_post_send/recv or mlx4_post_send/recv).

Why we should beware of the problem of backward compatibility for those variables?
mlx5/mlx4 can remain thread-safe, we don't rely on this behavior on OFI provider level.

Reply:
Anyway if libibverbs can tweak this behavior somehow, we can use this standard way.

Reply:
We're assuming that the libmlx4/mlx5 libraries under libibverbs support this environment variable.  How do we know that they do?  What versions of those libraries are required?  Do we have guarantees that future versions of the libraries will also support them?  Mellanox hasn't been the best about ensuring that backwards compatibility is maintained.  If we're dealing with a library version that doesn't have single-thread support, then we'll end up broken, but in very difficult to detect ways.

Reply:
To add to the comment on Parent Domains (PaD), Verbs has already standardized it. And also Thread Domains (TD), which is what the PaD contains. Currently, the TD only removes locks on the mlx5 hardware resources and not on the QP. (1) The patch for removing locks on the QP when it is assigned to a TD is being reviewed. To disable locks on the CQ through code, however, we will need to use the extended CQ (using it has its own caveats though). Extending the regular CQ to take in a PaD will break the ABI.

Reply:
@rzambre @shefty Thanks for educating me.
@shefty Agree, we can't know how does this variable behave in some version of libmlx5.